src/expr/src/relation/join_input_mapper.rs

```diff
@@ -0,0 +1,356 @@
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::ops::Range;
+
+use crate::RelationExpr;
+use crate::ScalarExpr;
+
+/// Any column in a join expression exists in two contexts:
+/// 1) It has a position relative to the result of the join (global)
+/// 2) It has a position relative to the specific input it came from (local)
+/// This utility focuses on taking expressions that are in terms of
+/// the local input and re-expressing them in global terms and vice versa.
+#[derive(Debug)]
+pub struct JoinInputMapper {
+    /// The number of columns per input. All other fields in this struct are
+    /// derived using the information in this field.
+    arities: Vec<usize>,
+    /// Looks up which input each column belongs to. Derived from `arities`.
+    /// Stored as a field to avoid recomputation.
+    input_relation: Vec<usize>,
+    /// The sum of the arities of the previous inputs in the join. Derived from
+    /// `arities`. Stored as a field to avoid recomputation.
+    prior_arities: Vec<usize>,
+}
+
+impl JoinInputMapper {
+    /// Creates a new `JoinInputMapper` and calculates the mapping of global context
+    /// columns to local context columns.
+    pub fn new(inputs: &[RelationExpr]) -> Self {
+        let types = inputs.iter().map(|i| i.typ()).collect::<Vec<_>>();
+        let arities = types
+            .iter()
+            .map(|t| t.column_types.len())
+            .collect::<Vec<_>>();
+
+        let mut offset = 0;
+        let mut prior_arities = Vec::new();
+        for input in 0..inputs.len() {
+            prior_arities.push(offset);
+            offset += arities[input];
+        }
+
+        let input_relation = arities
+            .iter()
+            .enumerate()
+            .flat_map(|(r, a)| std::iter::repeat(r).take(*a))
+            .collect::<Vec<_>>();
+
+        JoinInputMapper {
+            arities,
+            input_relation,
+            prior_arities,
+        }
+    }
+
+    /// All column numbers in order for a particular input in the local context
+    #[inline]
+    pub fn local_columns(&self, index: usize) -> Range<usize> {
+        0..self.arities[index]
+    }
+
+    /// All column numbers in order for a particular input in the global context
+    #[inline]
+    pub fn global_columns(&self, index: usize) -> Range<usize> {
+        self.prior_arities[index]..(self.prior_arities[index] + self.arities[index])
+    }
+
+    /// Takes an expression from the global context and creates a new version
+    /// where column references have been remapped to the local context.
+    /// Assumes that all columns in `expr` are from the same input.
+    pub fn map_expr_to_local(&self, expr: &ScalarExpr) -> ScalarExpr {
+        let mut expr = expr.clone();
```

Comment:
Very minor, but I thought I'd call out: in Rust idioms we might want a different signature, as this one forces a clone of the input. If instead an owned input was taken, the caller could decide if a clone is appropriate. It's not a huge deal at the moment, but I wanted to call out idiomatic Rust stuff like that!